src/polycap.h

```diff
@@ -0,0 +1,105 @@
+  typedef easy_rng polycap_rng;
+  #define polycap_rng_alloc(T) easy_rng_alloc(T)
+  #define polycap_rng_free(rng) easy_rng_free(rng)
+  #define polycap_rng_set(rng, seed) easy_rng_set(rng, seed)
+  #define polycap_rng_uniform(rng) easy_rng_uniform(rng)
+  #define polycap_rng_mt19937 easy_rng_mt19937
+#else
+  #include <gsl/gsl_rng.h>
+  #include <gsl/gsl_randist.h>
+  typedef gsl_rng_type polycap_rng_type;
+  typedef gsl_rng polycap_rng;
+  #define polycap_rng_alloc(T) gsl_rng_alloc(T)
+  #define polycap_rng_free(rng) gsl_rng_free(rng)
+  #define polycap_rng_set(rng, seed) gsl_rng_set(rng, seed)
+  #define polycap_rng_uniform(rng) gsl_rng_uniform(rng)
+  #define polycap_rng_mt19937 gsl_rng_mt19937
+#endif
+#include <complex.h> //complex numbers required for Fresnel equation (reflect)
+#include <gsl/gsl_multifit.h>
+#include <stdbool.h>
+
+// Structure definitions
+struct inp_file
+  {
+  double sig_rough;
+  double sig_wave;
+  double corr_length;
+  double d_source;
+  double d_screen;
+  double src_x;
+  double src_y;
+  double src_sigx;
+  double src_sigy;
+  double src_shiftx;
+  double src_shifty;
+  int nelem;
+  int iz[NELEM];
+  double wi[NELEM];
+  double density;
+  double e_start;
+  double e_final;
+  double delta_e;
+  int ndet;
+  int shape;
+  char prf[80];
+  char axs[80];
+  char ext[80];
+  double length; //in cm
+  double rad_ext[2]; //PC external radius, in cm
+  double rad_int[2]; //single capillary radius, in cm
+  double focal_dist[2]; //focal distance at both sides of PC, in cm
+  double n_chan;
+  char out[80];
+  };
+
+struct cap_prof_arrays
+  {
+  double zarr; //length axis (0 -> capillary length)
+  double profil; //single capillary shape (usually conical)
+  double sx; //polycapillary central axis x coordinates
+  double sy; //polycapillary central axis y coordinates
+  double d_arr; //external polycapillary shape
+  };
+
+struct cap_profile
+  {
+  int nmax; /*nr of points defined along capillary profile*/
+  double rtot1; /*radius at start position*/
+  double rtot2; /*radius at end position*/
+  double cl;    /*capillary length*/
+  double binsize; /*20.e-4 cm*/
+  struct cap_prof_arrays *arr; /* will get proper size allocated to it later, typically 1000*sizeof(struct cap_prof_arrays) */
+  };
+
+// Functions
+struct mumc *ini_mumc(struct inp_file *cap) // Calculate total cross sections and scatter factor
+int segment(double s0[3], double s1[3], double rad0, double rad1, double rh1[3], double v[3], double rn[3], double *calf) // calculates the intersection point coordinates of the photon trajectory and a given linear segment of the capillary wall
+double polycap_refl(double e, double theta, double density, double scatf, double lin_abs_coeff) //calculates reflectivity according to Fresnel equation
+struct cap_profile *def_cap_profile(unsigned long int shape, double length, double rad_ext[2], double rad_int[2], double focal_dist[2]) //calculates polycapillary shapes (shape 0: cone, 1: paraboloid, 2: ellipsoid)
+
+
```

Comment:
@tschoonj Is dit oke als header file, of sla ik de bal compleet mis? (kijkende naar xmimsim vind ik sommige header files die er zo uitzien, andere zijn een pak compacter...)

Reply:
Zie mijn andere comments 😄 

Je moet trouwens niet alles in 1 header steken. Beter is te werken met publieke en private headers: publieke bevat enkel wat andere mensen die uw library will gebruiken in hun software nodig hebben en wordt dus geïnstalleerd. Private headers daarentegen bevatten functies en struct implementaties die nuttig zijn voor jouw development en voor de unit tests te laten werken en zijn uiteraard deel van de source code maar worden niet geïnstalleerd.

Reply:
Oke, confusion strikes again :P 

a) publieke en private headers: is niet alles publiek daar het op github staat? Concreet, in dit geval, lijkt het me dat anderen vooral het programma als geheel zouden gebruiken (de verschillende onderdelen zijn niet zo nuttig op zich). Wat eventueel nuttig kan zijn is dat ze bijvoorbeeld de Fresnel vergelijking kunnen laten berekenen in hun eigen programma, of ook een polycapillary shape maken etc, dus daarom had ik de functies die daarmee overeenkomen in de header gezet. Dit is dan publiek? Hoe verschilt de private header dan van een regelrechte kopie van de eerste 160 lijnen in polycap.c?
Alles wat bovendien nu in de header staat, staat ook nog eens in de .c file. Dat is toch de bedoeling, juist? (dacht dat je het zo had gezegd toen je hier was)

b) de structures die ik momenteel in polycap.h had gezet was zodat de gebruiker zou inzien hoe die structure eruit ziet. Immers als hij de functies wil oproepen moet toch een gelijke structure voorzien worden? Ik vermoed dat het dat is wat je bedoelt met "aan de inhoud van je structs kunnen". In dat geval denk ik dus dat de gebruikers wel aan de inhoud van de structs moeten kunnen, daar ze toch een input moeten kunnen genereren die voor de functies steek houdt?

c) Zal idd nog eens kijken om nog wat zaken dynamisch te laten alloceren. NELEM wordt nog gedefinieerd in polycap.c. Dit moet dan ook in polycap.h staan? Kan het in dit geval wel dynamisch laten alloceren denk ik, maar kvroeg het me gewoon af.
Ik moet dus in principe ook in polycap.h alles met #define zetten wat gebruikt wordt in de 'publieke functies' dan? Maar de andere #defines, die enkel gebruikt worden in main code (en dus niet die kleinere functies die mss ook handig zijn voor gebruikers op zich) zet ik niet in polycap.h. En voor de zekerheid, alles wat ik in polycap.h staat, staat ook nog eens in polycap.c he?

Sorry voor de vele vragen hoor... In mijn hoofd is een header file gewoon die eerste 160 lijnen van polycap.c in het algemeen, daar dat is wat de code nodig heeft om te kunnen draaien...

Reply:
De wikipagina over de include guard heeft al een ander duidelijker gemaakt denk ik.
Het is dus de bedoeling dat ik belangrijke zaken in polycap.h zet (publiek), met een include guard, en dan in polycap.c de overeenkomstige stukken weglaat en vervang door #include "polycap.h"?
Zodat iemand anders dan gemakkelijk polycap.h kan gebruiken in hun code, maar polycap.c nog wat extras heeft (private, declared in polycap.c?)

Moeten dan mijn structures in polycap.h niet sowieso expliciet defined worden? Anders weten de users toch van geen kanten welke onderdelen de structure die ze willen meegeven aan een functie moet bevatten?

Reply:
a) Het verschil tussen publieke en private headers ligt erin welke geïnstalleerd wordt en dus toegankelijk is voor third-parties die in hun software gebruik willen maken van polycap. In de publieke steek je dus het absolute minimum dat ze dus nodig hebben om te kunnen werken maar ook niet meer dan dat. Private headers bevatten functies uit je shared library die intern wil houden, en die nodig zijn bijvoorbeeld voor je unit tests als ze die zouden willen testen, of indien je je source code zou opsplitsen over meerdere bestanden: de private headers zijn dan nodig om te weten wat de signature is van een functie uit een ander bestand.
Nog een belangrijk verschil tussen privaat en publiek is dat je private headers zoveel mag aanpassen als je wil zonder je zorgen te maken over backwards compatibility. Als je daarentegen functies verwijdert of de signature verandert in publieke headers zal alle software die gelinkt geweest is tegen een oudere versie niet meer werken wanneer de nieuwe polycap library wordt geïnstalleerd: in dat geval moet de third party zijn source code aanpassen rekening houdende met de wijzigingen en hercompileren. Je moet je ook geen zorgen maken over het feit dat private headers gewoon in de source code zitten en zichtbaar zijn voor iedereen. In theorie kan de third party die gaan gebruiken om directe toegang te hebben tot de member variables van structs maar dat zou bijzonder lomp zijn omdat jij als polycap developer die naar genoegen kan gaan wijzigen, waardoor de third party zijn software waarschijnlijk zal segfaulten wanneer een nieuwe versie van polycap wordt geïnstalleerd.

b) Enkel als je wil dat de third party directe toegang heeft tot de inhoud van de structs moet je ze in de publieke header zetten anders niet. Als je dat doet kan je trouwens ook geen wijzigingen aan de struct aanbrengen zonder opnieuw backwards compatibility in gevaar te brengen. Als jij ineens bijvoorbeeld beslist om een struct member te verwijderen dat gebruikt werd door de third party, zal die waarschijnlijk een segfault krijgen indien de nieuwe versie van de library gebruikt wordt. Best is structs met hun volledige definitie te zetten in private headers zodat ze kunnen gebruikt worden vanuit alle bestanden in jouw source.
In de praktijk wordt dit probleem opgelost door te werken met pointers naar structs, gecombineerd met getter en setter functies voor die structs zodanig dat de third party toch info kan bekomen over de structs, terwijl de inhoud toch afgeschermd is van de third party. Setter functies hebben ook het voordeel dat de nieuwe waarde voor een member variabele eerst kan gecontroleerd worden. Een voorbeeld:

Publieke header (`data_public.h`):
````c
struct myData;

struct myData* data_new(int value);
void data_set_value(struct myData *my_data, int new_value);
int data_get_value(struct myData *my_data);
void data_free(struct myData *my_data);
````

Private header (`data_private.h`):
````c
struct myData {
  int value;
};
````

Implementatie (`data.c`):
````c
#include "data_public.h"
#include "data_private.h"

struct myData* data_new(int value) {
  struct myData *data = malloc(sizeof(struct myData));
  data->value = value;
  return data;
}
void data_set_value(struct myData *my_data, int new_value) {
  my_data->value = new_value;
}

int data_get_value(struct myData *my_data) {
  return my_data->value;
}

void data_free(struct myData *my_data) {
  free(my_data);
}
````

c) Zoals het nu is zou de header niet werken, want `NELEM` is niet gedefinieerd. De third party zou enkel een compiler error krijgen als hij die header zou willen gebruiken. Publieke headers moeten zelf consistent zijn: iedereen moet ze kunnen gebruiken zonder andere headers te importeren of macros te gaan definiëren.

Wat je zeker niet mag doen is wat in je publieke of private headers staat te gaan dupliceren in de implementatie! Niet alleen is dat nutteloos want het kan vermeden worden met een eenvoudige include, maar het betekent ook dat elke verandering in de header ook moet gedaan worden in de implementatie...

Maak je geen zorgen over de vragen: ik help u met veel plezier!

Reply:
Nog een voordeel van het afschermen van struct member variables en het gebruik van getters en setters (eigenlijk een soort van lightweight object oriented programmeren), is dat je dan makkelijk bindings kan schrijven voor Fortran en Python, wanneer de tijd rijp is daarvoor 😄